### ISSUE TITLE
Error when deploy on MacOS 10.13.3

### ISSUE THREAD

**Issue by thoangtd** [kangaroo-server/kangaroo#428]
Scenario:
Enable logger:  
<logger name="org.hibernate.orm.deprecation" level="ON"/>
 <logger name="org.reflections.Reflections" level="ON"/>

Build project with:  mvn clean install -P mariadb -DskipTests=true
Deploy : sudo java -jar target/kangaroo-server-authz-1.2.3-SNAPSHOT-jar-with-dependencies.jar
It through exceptions : 
[error-log.txt](https://github.com/kangaroo-server/kangaroo/files/1697638/error-log.txt)

Environment: 
java version "1.8.0_112"
Java(TM) SE Runtime Environment (build 1.8.0_112-b16)
Java HotSpot(TM) 64-Bit Server VM (build 25.112-b16, mixed mode)

OS: MacOS 
Maven: 3.5.2

**Comment by krotscheck** [comment]
So, I suspect you're running on an older version of `kangaroo-common`, as what you're seeing is just a noisy logger from the reflections API, fixed in this patch: https://github.com/kangaroo-server/kangaroo/pull/426

Also, a few pointers (which, I know, are not in the documentation and I really should update that:

- The `-P mariadb` flag is really only meaningful in testing, as it switches out some system parameters used in the surefire runner.
- Running the server with `sudo` is interesting, are you attaching the server to port 443 directly?

If you'd like to use the published jar directly instead of compiling your own, you can switch the database directly, as all hibernate parameters are exposed via the commandline API. In other words:

```
java -jar kangaroo-server-authz-1.2.2-jar-with-dependencies.jar \
   -Dhibernate.connection.url=mysql://localhost:3306/my_database 
   -Dhibernate.connection.username=user /
   -Dhibernate.connection.password=root /
   -Dhibernate.connection.driver_class=org.mariadb.jdbc.Driver /
   -Dhibernate.dialect=org.hibernate.dialect.MariaDBDialect
```

Will get you the same results.

**Comment by thoangtd** [comment]
I had get lasted version. It through exception when turn on log
  <logger name="org.hibernate.orm.deprecation" level="ON"/>
  <logger name="org.reflections.Reflections" level="ON"/>
Because I want to see log on server when it response exception
{
    "error": "bad_request",
    "error_description": "HTTP 400 Bad Request"
}
In the currently it's not logger out.
I use "postman" tool try to update application. It's work fine with GET method, but when using "PUT/POST", it response "Bad request" but i don't see error in the logger. pls tell me how to config logger in this case.
<img width="1214" alt="screen shot 2018-02-07 at 11 47 03" src="https://user-images.githubusercontent.com/2314216/35899037-b401f7f6-0bfc-11e8-9228-0ba6f66f066e.png">
<img width="1231" alt="screen shot 2018-02-07 at 11 46 12" src="https://user-images.githubusercontent.com/2314216/35899038-b4321134-0bfc-11e8-9fc4-4acd0b35c361.png">

I had put Sysout but it seem not run to here 
` @PUT
    @Path("/{id: [a-f0-9]{32}}")
    @Consumes(MediaType.APPLICATION_JSON)
    @Produces(MediaType.APPLICATION_JSON)
    @ApiOperation(value = "Update application")
    public Response updateResource(
            @io.swagger.annotations.ApiParam(type = "string")
            @PathParam("id") final BigInteger id,
            final Application application) {
        System.out.println("Update application");
        Session s = getSession();

        // Load the old instance.
        Application currentApp = s.get(Application.class, id);

        assertCanAccess(currentApp, getAdminScope());

        // Additional special case - we cannot modify the kangaroo app itself.
        if (currentApp.equals(getAdminApplication())) {
            throw new ForbiddenException();
        }

        // Make sure the body ID's match
        if (!currentApp.equals(application)) {
            throw new BadRequestException();
        }

        // Make sure we're not trying to change data we're not allowed.
        if (!currentApp.getOwner().equals(application.getOwner())) {
            throw new BadRequestException();
        }

        // Did the role change?
        if (!Objects.equals(currentApp.getDefaultRole(),
                application.getDefaultRole())) {

            // Can't null it if it's already been set.
            if (application.getDefaultRole() == null) {
                throw new BadRequestException();
            }

            // Make sure the new role belongs to this application.
            Role resolveDesired =
                    s.get(Role.class, application.getDefaultRole().getId());
            if (resolveDesired == null || !Objects.equals(
                    resolveDesired.getApplication(), application)) {
                throw new BadRequestException();
            }
        }

        // Transfer all the values we're allowed to edit.
        currentApp.setName(application.getName());

        s.update(currentApp);

        return Response.ok(application).build();
    }`

**Comment by krotscheck** [comment]
Ok, that's a lot in one post. Let me split this up into individual tickets and let's take the discussion there.

**Comment by krotscheck** [comment]
(To get you unblocked on the API request though: The Bad Request is likely thrown by the `CSRFProtectionFilter`. If you don't include a `X-Requested-With` header it'll just reject the response, as per the OWASP CSRF Protections. I can add logging to make that explicit.).

**Comment by krotscheck** [comment]
Alright, I've created the various issues. I'm really curious about your use case, because so far I've been working only to support my own. Could you perhaps give me some idea on what features you're looking for / trying to evaluate this for? It'd be pretty quick for me to read through your requirements and give you an idea of whether Kangaroo supports them, and how much additional work any gaps might require.

**Comment by thoangtd** [comment]
Thanks for support.
In my case, I want re-implement database layer, because my system had managed apps, client, user ... 
example: i used mongodb for storage database apps, client, user... i will CRUD from my database to providing oauth-server.I think It will be great, you can support define entity and service interface for CRUD database.

**Comment by krotscheck** [comment]
Alright, if I understand this correctly, you'd like to consume the majority of the code, however you already have an existing database layer that you'd like to use, and it happens to be MongoDB.

Off the top of my head, I think this would require three major bodies of work:
- Step 1: Disconnect from hibernate explicitly and only use JPA and JTA annotations: https://github.com/kangaroo-server/kangaroo/issues/250
- Step 2: Move all of our database foreign key constraints into JPA annotations, so that the ORM becomes responsible for foreign key validation.
- Step 3: With those two pieces in place, you should be able to consume the kangaroo-authz-server as a jar directly, and assemble your own server using the `ServerFactory` and a feature collector like `AdminV1API`, omitting the HibernateFeature and the DatabaseFeature and replacing them both with your own `Feature` for MongoDB.

I do not mind doing steps 1 and 2, though they're not the highest priority right now. I'll look into them after I finish the initial draft of the User Interface :)
